docs/proposals/shoot-control-plane-migration/README.md

```diff
@@ -0,0 +1,234 @@
+
+There is also a future possibility of having a set of seed clusters serving a particular set of shoot clusters on a cloud provider, region etc. to address both reliability as well as load distribution among other issues.
+
+Given such challenges and requirements, we need some mechanism to migrate/move a shoot cluster's control plane between seed clusters.
+
+## Goal
+
+* Provide a mechanism to migrate/move a shoot cluster's control-plane between seed clusters.
+* The mechanism should make it possible to reuse the shoot cluster nodes that are already available and healthy. Such existing available and healthy nodes should be reused after the control-plane is moved to the destination seed cluster.
+* The mechanism should be extensible in the future to support migration across regions/availability zones.
+* The mechanism should be automated.
+* The mechanism should work regardless whether the source seed cluster and the shoot control-plane running there is available and healthy or not. I.e., the mechanism should work for the disaster recovery scenario as well as a scenarios such as seed load-balancing.
+
+To be more explicit, the following scenarios should be supported.
+
+| Scenario | Source Seed Reachable from Garden Cluster | Source Seed Running | Source Seed Can Reach Anything Outside |
+| ---:| --- | --- | --- |
+| 1 | Yes | Yes/No | Yes/No |
+| 2 | No | Yes | Yes |
+| 3 | No | No | No |
+| 4 | No | Yes | No |
+
+## Non-goal
+
+* The proposed mechanism need not right away implement a solution for migration across regions/availability zones.
+
+## Reuse
+
+* [Machine Controller Manager](https://github.com/gardener/machine-controller-manager) already has the capability to adopt existing machines in the IaaS layer based on proper machine definitions.
+* [Etcd Backup Restore](https://github.com/gardener/etcd-backup-restore) already backs up the shoot etcd which is running in the shoot namespace of the source seed cluster. We can re-use this completely while restoring the shoot etcd on the destination seed cluster.
+
+## Proposed Solution
+
+### Backup and Restoration
+
+The backup and restoration of the following resources that are currently not stored in the shoot apiserver/etcd is not covered by the [reuse](#reuse) mentioned above.
+
+* Resources from the shoot namespace in the seed.
+  * Terraform resources such as configmaps and secrets.
+  * Machine Controller Manager resources such as machinedeployments, machinesets, machines, machineclasses, secrets etc.
+* Resources from the shoot backup namespace in the seed.
+  * Terraform resources such as configmaps and secrets.
+
+All the additional resources listed above will be backed up in the same object store as the what is used for backing up the shoot etcd.
+
+* The backup will be watch-based and not schedule based.
+* The backup will be full backups at longer intervals followed by incremental backups at smaller intervals similar to the etcd continuous/incremental backup.
```

Comment:
Yup. It will be file for file(resource object YAML) implementation. But when we say incremental, only file/object for which watch event triggered, only files associated with it and objects on which it depends will be updated with new copy. 

>  backup will be full backups at __longer interval__ ...

I forgot to update it, @amshuman-kr ,  in my opinion here concept of periodic backup dosen't make sense. We can rely on informer's watch functionality to not miss any event on object.

Reply:
Regarding not having periodic base backup, I am not sure. If we do not have it then the algorithm for finding the right files to restore should be optimised. A naive implementation replaying all the changes would be very sub-optimal in that case.

Reply:
Regarding file-for-file, I agree. But we need to support some way to record deletion. A naive file-for-file implementation will not support object/file deletion.

Reply:
Yes. Deletion needs some special logic, but probably it can be quite easily handled by some checks on deletionTimestmap or finalizer if any.